addons/misra.py

```diff
@@ -1579,6 +1580,49 @@ def misra_17_1(self, data):
             elif token.str == 'va_list':
                 self.reportError(token, 17, 1)
 
+    def misra_17_2(self, data):
+        def find_indirect_calls(func_id, calls_map, call_tokens_map, found_calls, visited):
+            for tok_id in calls_map.get(func_id, []):
+                tok = call_tokens_map[tok_id]
+                call_id = tok.function.Id
+                if call_id in visited:
+                    found_calls.add(tok_id)
+                calls_map.get(func_id).remove(tok_id)
+                visited.add(func_id)
+                found_calls.update(find_indirect_calls(call_id, calls_map, call_tokens_map, found_calls, visited))
+            return found_calls
+
+        # function id : list of token ids with its calls
+        function_calls = {f.Id: [] for f in data.functions}
+        call_tokens_map = {}
+        for scope in data.scopes:
+            if scope.type != 'Function':
+                continue
+
+            in_token = scope.bodyStart.next
+            while in_token != scope.bodyEnd and in_token is not None:
+                name_token = in_token.astOperand1
+                if isFunctionCall(in_token) and name_token.function:
+                    if name_token not in function_calls[scope.function.Id]:
```

Comment:
It seems weird to use the function Id as key. That is a string value. String comparisons will be used for the dict lookup here. Is it possible to have some "Pointer comparisons" instead?

Reply:
Hm, actually, no. I use [deepcopy](https://github.com/danmar/cppcheck/pull/1962/files#diff-1bfaaa2a6da0266311b76e7291e15236R1620) to pass those values into search function and [modify](https://github.com/danmar/cppcheck/pull/1962/files#diff-1bfaaa2a6da0266311b76e7291e15236R1590) this copies inside it. If I'll use a token values instead of strings, `deepcopy` operation will create copy for all lxml etree elements. This is very memory-expensive and also may cause recursion max depth error, because it will recursively copy all tree elements.

Reply:
ok I see your point.. hmm I am not convinced deepcopy should be used.